system-addon/lib/DownloadsManager.jsm

```diff
@@ -98,17 +96,31 @@ this.DownloadsManager = class DownloadsManager {
     }
   }
 
-  get downloads() {
+  async getDownloads({threshold = 0, numItems = this._viewableDownloadItems.size, onlyExists = false}) {
     let results = [];
-    for (const url of this._viewableDownloadItems.keys()) {
-      const elem = this._viewableDownloadItems.get(url);
-      // We only want downloads that are less than 36 hours old
-      if (elem && (elem.downloadedSince < RECENT_DOWNLOAD_THRESHOLD)) {
-        const formattedDownloadForHighlights = this.formatDownload(elem);
-        results.push(formattedDownloadForHighlights);
+    for (const elem of this._viewableDownloadItems.values()) {
+      // Only get downloads within the time threshold specified
+      if (elem && (elem.downloadedSince < threshold)) {
+        // If we only care about downloads where the file still exists, refresh them
+        // to ensure the 'exists' attribute is up to date
+        if (onlyExists) {
+          await elem.download.refresh();
```

Comment:
As discussed on irc, I'm assuming `refresh`, which results in file IO, is significantly slower than anything else we do in this method whether that's iterating or sorting, so we want to minimize the number of times we call it. This can be done by only checking the first few most recent downloads in order, so if the first one exists, we're done.